Add item requests

LeonovichMariia opened the pull request.

LeonovichMariia committed:
feat: добавлены классы: контроллер, сервис и хранилище для ItemRequest. Реализована логика  добавления нового запроса вещи и получения различных списков запросов.
Реализована функциональность ответа на запрос.
Добавлена пагинация для : GET /requests/all, GET /items, GET /items/search, GET /bookings и GET /bookings/owner.

LeonovichMariia committed:
fix: code style

LeonovichMariia committed:
fix: в классы AddItemRequest, ItemRequestCreationRequest добавлено поле requestor. Внесены соответствующие правки в мапперы и в логику.
delete: из классов AddItemRequest, ItemRequestCreationRequest удалены поля name, requestId, available.
feat: добавлен тестовый класс ItemRequestControllerTest.

LeonovichMariia committed:
delete: удален класс ItemRequestCreationRequest, вместо него используется AddItemRequest.
feat: в класс ItemRequestControllerTest добавлены тесты для REST-эндпоинтов.

LeonovichMariia committed:
feat: добавлены юнит-тесты, интеграционные тесты, тесты репозиториев и тесты для DTO.

LeonovichMariia committed:
feat: добавлены еще тесты для более полного покрытия кода

LeonovichMariia committed:
delete: удален тест addBookingWithEqualStartAndEndTime;

KirPyt commented on src/main/java/ru/practicum/shareit/booking/controller/BookingController.java:
```diff
@@ -46,15 +51,19 @@ public BookingDto getBookingById(@RequestHeader("X-Sharer-User-Id") Long userId,
-                                               @RequestParam(defaultValue = "ALL") BookingState state) {
+                                               @RequestParam(defaultValue = "ALL") BookingState state,
+                                               @RequestParam(name = "from", defaultValue = "0") @PositiveOrZero Integer from,
+                                               @RequestParam(name = "size", defaultValue = "10") @Min(1) @Max(20) Integer size) {
```
Ограничений сверху на количество элементов согласно ТЗ нет

KirPyt commented on src/main/java/ru/practicum/shareit/booking/controller/BookingController.java:
```diff
@@ -46,15 +51,19 @@ public BookingDto getBookingById(@RequestHeader("X-Sharer-User-Id") Long userId,
-                                              @RequestParam(defaultValue = "ALL") BookingState state) {
+                                                    @RequestParam(defaultValue = "ALL") BookingState state,
+                                                    @RequestParam(name = "from", defaultValue = "0") @PositiveOrZero Integer from,
+                                                    @RequestParam(name = "size", defaultValue = "10") @Min(1) @Max(20) Integer size) {
```
Аналогично, ограничений сверху на `size` нет

KirPyt commented on src/main/java/ru/practicum/shareit/booking/service/BookingServiceImpl.java:
```diff
@@ -90,28 +91,29 @@ public BookingDto getBookingById(Long bookingId, Long userId) {
         userRepository.validateUser(bookerId);
         LocalDateTime now = LocalDateTime.now();
         List<Booking> bookingDtoList;
+        PageRequest pageable = PageRequest.of(from > 0 ? from / size : 0, size, SORT);
```
Рекомендую подумать над тем, чтобы сформировать класс, наследуемый от `PageRequest` и определить для него конструктор, принимающий параметры `from` и `size`. Данный класс будет реализовывать логику расчёта пагинации. Таким образом алгоритм расчёта пагинации будет инкапсулирован в данный класс

KirPyt commented on src/main/java/ru/practicum/shareit/booking/service/BookingServiceImpl.java:
```diff
@@ -123,28 +125,31 @@ public List<BookingDto> getAllUserBookings(Long bookerId, BookingState state) {
         userRepository.validateUser(userId);
         LocalDateTime now = LocalDateTime.now();
         List<Booking> bookingDtoList;
+        PageRequest pageable = PageRequest.of(from > 0 ? from / size : 0, size, SORT);
```
Согласно валидации в контроллере `from` не может быть меньше 0, поэтому данная проверка лишняя
В методе выше аналогично

KirPyt commented on src/main/java/ru/practicum/shareit/item/controller/ItemController.java:
```diff
@@ -45,15 +49,19 @@ public ItemDto getItemById(@RequestHeader("X-Sharer-User-Id") Long ownerId,
-    public List<ItemDto> getPersonalItems(@RequestHeader("X-Sharer-User-Id") Long userId) {
+    public List<ItemDto> getPersonalItems(@RequestHeader("X-Sharer-User-Id") Long userId,
+                                          @RequestParam(name = "from", defaultValue = "0") @Min(0) Integer from,
+                                          @RequestParam(name = "size", defaultValue = "10") @Min(1) @Max(20) Integer size) {
```
Здесь и далее аналогичное замечание про `@Max(20)`

KirPyt commented on src/main/java/ru/practicum/shareit/item/dto/ItemDto.java:
```diff
@@ -25,7 +24,7 @@ public class ItemDto {
     @NotNull(message = ValidationMessages.AVAILABLE_NULL)
     private Boolean available;
     private User owner;
```
В ДТО-шках не должно быть конкретных сущностей, только `id`

KirPyt commented on src/main/java/ru/practicum/shareit/request/dto/AddItemRequest.java:
```diff
@@ -1,21 +1,32 @@
+@Builder
+@NoArgsConstructor
+@AllArgsConstructor
+public class AddItemRequest {
```
Рекомендую добавить в название класса `Dto`, чтобы придерживаться единообразия наименования ДТО-шек

KirPyt commented on src/main/java/ru/practicum/shareit/request/dto/ItemRequestMapper.java:
```diff
@@ -0,0 +1,35 @@
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class ItemRequestMapper {
```
В данный маппер по аналогии с остальными мапперами можно добавить `@UtilitiClass`

KirPyt commented on src/test/java/ru/practicum/shareit/item/repository/CommentRepositoryTest.java:
```diff
@@ -0,0 +1,124 @@
+@DataJpaTest
+class CommentRepositoryTest {
+    @Autowired
+    CommentRepository commentRepository;
```
Поля надо инкапсулировать, а так же рекомендую подумать и о финализации полей

KirPyt commented on src/test/java/ru/practicum/shareit/user/service/UserServiceImplTest.java:
```diff
@@ -0,0 +1,104 @@
+    @Mock
+    private UserRepository userRepository;
+    @InjectMocks
+    UserServiceImpl userService;
```
Надо инкапсулировать

KirPyt reviewed (commented):
Всё замечательно, тесты отлично написаны. Оставил некоторые незначительные замечания по коду, ну и советы так же. Предлагаю поработать с ними.

Если будут вопросы, то пишите в Пачку

LeonovichMariia committed:
delete: удалена аннотация @Max(), удалена лишняя проверка параметра from в методах сервисов.
fix: в классе ItemDto тип данных поля owner изменено с User на Long. В название класса AddItemRequest добавлено Dto. В класс ItemRequestMapper добавлена аннотация @UtilityClass. В классах CommentRepositoryTest и UserServiceImplTest инкапсулированы поля.
feat: создан класс PageSetup, реализующий логику расчёта пагинации.

KirPyt commented on src/test/java/ru/practicum/shareit/request/controller/ItemRequestControllerTest.java:
```diff
@@ -0,0 +1,173 @@
+    @MockBean
+    private ItemRequestService itemRequestService;
+    private AddItemRequestDto itemRequest;
+    DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss.SSS");
```
Не забывайте про ключевые слова - здесь отличным решением было бы сочетание `final` и `static`

KirPyt commented on src/test/java/ru/practicum/shareit/request/service/ItemRequestServiceImplTest.java:
```diff
@@ -0,0 +1,144 @@
+    @Mock
+    private UserService userService;
+    @InjectMocks
+    ItemRequestServiceImpl itemRequestService;
```
Инкапсуляция

KirPyt reviewed (commented):
Очень хорошо!
Ваш проект зачтён

LeonovichMariia committed:
fix: переменные типа DateTimeFormatter сделаны static и final. В классе ItemRequestServiceImplTest инкапсулировано поле itemRequestService.
